services/base.js

```diff
@@ -306,6 +308,45 @@ class BaseService {
       })
     )
   }
+
+  static _validate(data, schema) {
+    if (!schema || !schema.isJoi) {
+      throw Error('A Joi schema is required')
+    }
+    const { error, value } = Joi.validate(data, schema, {
+      allowUnknown: true,
+      stripUnknown: true,
+    })
+    if (error) {
+      trace.logTrace(
+        'validate',
+        emojic.womanShrugging,
+        'Response did not match schema',
+        error.message
+      )
+      throw new InvalidResponse({
+        prettyMessage: 'invalid data response',
+        underlyingError: error,
+      })
+    } else {
+      trace.logTrace(
+        'validate',
+        emojic.bathtub,
+        'Data after validation',
+        value,
+        { deep: true }
+      )
+      return value
+    }
+  }
+
+  async _request({ url, options = {}, errorMessages = {} }) {
+    const logTrace = (...args) => trace.logTrace('fetch', ...args)
+    logTrace(emojic.bowAndArrow, 'Request', url, '\n', options)
+    const { res, buffer } = await this._sendAndCacheRequest(url, options)
+    logTrace(emojic.dart, 'Response status code', res.statusCode)
+    return checkErrorResponse.asPromise(errorMessages)({ buffer, res })
```

Comment:
AFAICT the only difference between json and xml is the parsing – is that right?

Another way we could consider handling the relationship between `_request` and `_validate` is to have `_request` delegate to a `_parse` function, and put the format-specific bits there. The service code would invoke `_request` which would be BaseService's implementation, and that would in turn invoke `_parse` which would be BaseXxService's.

This would force me to change slightly the way I handled #2031 because every service would then need a `_parse` implementation. It would also generally force everything to be a native object that can be checked by a schema. However I don't see a problem with that. We could make a BaseSvgBadge subclass easily enough. And encouraging schema validation is good.

I like this idea because the `_validate` function is in BaseService so it would feel natural to call it from its `_request` function. The logging code probably could be moved from BaseXxService to BaseService as well.

Reply:
The requests are slightly different as weel, in particular the `Accept` headers. If we go with your suggestion, we'll also have to add a `get defaultHeaders` function or something like that, which we would override in the implementations.

Reply:
Ah, right, the request options, that's a good point. The headers could also be added in a `_request` override. Though the differences in the requests may mean it's clearer to handle this the way it is.